Accessing Kafka: Part 4 - Load Balancers (#64)
strimzi/strimzi.github.io: _posts/2019-05-06-accessing-kafka-part-4.md

---
layout: post
title:  "Accessing Kafka: Part 4 - Load Balancers"
date: 2019-05-06
author: jakub_scholz
---

In the fourth part of this blog post series we will look at exposing Kafka using load balancers.
This post will explain how to use load balancers in public cloud environments and how they can be used with Kafka.

<!--more-->

_This post is part of a bigger series about different ways to access a Kafka cluster powered by Strimzi.
The other parts published so far are:_

* _[Part 1 - Introduction](https://strimzi.io/2019/04/17/accessing-kafka-part-1.html)_
* _[Part 2 - Node Ports](https://strimzi.io/2019/04/23/accessing-kafka-part-2.html)_
* _[Part 3 - OpenShift Routes](https://strimzi.io/2019/04/30/accessing-kafka-part-3.html)_
* _Part 4 - Load Balancers (this post)_

# Load balancers

Load balancers automatically distribute incoming traffic across multiple targets.

> Different implementations do the traffic distribution on different level:

Reviewer: Not sure but should it be plural here? "levels" I mean @tombentley

> * Layer 7 load balancers can distribute the traffic on the level of individual requests (for example HTTP requests)

Reviewer: Adding "at application level"? Referring to ISO/OSI

* Layer 4 load balancers distribute the TCP connections

Load balancers are available in most public and private clouds.
Examples of load balancers are Elastic Load Balancing services from Amazon AWS, Azure Load Balancer in Microsoft Azure public cloud or Google Cloud Load Balancing service from Google.

> Load balancing services are also available in OpenStack.

Author: I do not know much about OpenStack. But I will add a link to openStack docs which talk about load balancers as a service.

If you run your Kubernetes or OpenShift cluster on bare metal, you might not have load balancers available on demand.
In that case, using node ports, OpenShift Routes or Ingress might be a better option for you.

Do not get scared by the long list of different load balancing services.
Most of them are very well integrated with Kubernetes.

> When the Kubernetes service as configured with the type `Loadbalancer`, Kubernetes will automatically create the load balancer through the cloud provider, which understands the different services offered by given cloud.

Reviewer: "is configured"

Thanks to that the Kubernetes applications - including Strimzi - do not need to understand the differences and should work everywhere where the cloud infrastructure and Kubernetes are properly integrated.

# Using load balancers in Strimzi

Since none of the common load balancing services supports the Kafka protocol, Strimzi always uses the Layer 4 load balancing.
Since Layer 4 works on the TCP level, it means that the load balancer will always take the whole TCP connection and direct it to one of the targets.
That has some advantages - you can for example decide whether TLS encryption should be enabled or disabled.

To give Kafka clients access to the individual brokers, Strimzi creates a separate service with `type=Loadbalancer` for each broker.

> As a result, each broker will get a separate load balancer _(despite the Kubernetes service being of a load balancer type, the load balancer is still a separate entity manages by the infrastructure / cloud)_.

Reviewer: "managed by..."

A Kafka cluster with N brokers will need N+1 load balancers.

![Accessing Kafka using load balancers]({{ "/assets/2019-05-06-per-pod-load-balancers.png" }})

On the beginning of this post we defined load balancer as something what _distribute incoming traffic across multiple targets_.
However, as you can set from the diagram above, the per-broker load balancers have only one target and are technically not load balancing.
That is true, but in most cases the actual implementation is a bit more complicated.

> When Kubernetes create the load balancer, they usually target it to all nodes of your Kubernetes cluster and not only to the nodes where your application is actually running.

Reviewer: creates

That means that although the TCP connections will always end on the same node in the same broker, they might be routed through the other nodes of your cluster.
When the connection is sent by the load balancer to the node which does not host the Kafka broker, the `kube-proxy` component of Kubernetes will forward it to the right node where the broker runs.

> This can lead to some delays since some fo the connections might be routed through more hops then absolutely necessary.

Reviewer: typo "of"

      # ...
      external:
        type: loadbalancer
    # ...
```

Similarly to the node port external listener, also with load balancers TLS is enabled by default.
If you don't want to use TLS encryption, you can easily disable it:

```yaml
    # ...
    listeners:
      external:
        type: loadbalancer
        tls: false
    # ...
```

After Strimzi creates the load balancer type Kubernetes services, the load balancers will be automatically created.
Most clouds will automatically assign the load balancer some DNS name and IP addresses.
These will be automatically propagated into the `status` section of the Kubernetes service.
Strimzi will read it from there and use it to configure the advertised address in the Kafka brokers.
When available, Strimzi currently always prefers the DNS name over the IP address.
The reason is that the IP addresses are often volatile while the DNS name is fixes for the whole lifetime of the load balancer (this applies at least to Amazon AWS ELB load balancers).
But if the load balancer has only an IP address, Strimzi will of course use it.

As a user, you should always use the bootstrap load balancer address for the initial connection.

> You can get the address form the `status` section with following command (replace `my-cluster` with the name of your cluster):

Reviewer: Typo "from"

```
kubectl get service my-cluster-kafka-external-bootstrap -o=jsonpath='{.status.loadBalancer.ingress[0].hostname}{"\n"}'
```

In case there is no hostname set, you can also try the IP address (replace `my-cluster` with the name of your cluster):

```
kubectl get service my-cluster-kafka-external-bootstrap -o=jsonpath='{.status.loadBalancer.ingress[0].ip}{"\n"}'
```

The DNS or IP returned by one of these commands can be used in your clients as the bootstrap address.
The load balancers use always the port `9094` to expose Apache Kafka.

> For more details, see the [Strimzi documentation](https://strimzi.io/docs/latest/full.html#proc-accessing-kafka-using-loadbalancers-deployment-configuration-kafka).

Author: I will add a similar CLI command (using the Kafka console producer) as I had in the previous part.

The following example shows the OpenStack annotations:

```yaml
apiVersion: kafka.strimzi.io/v1beta1
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:
    # ...
    template:
      externalBootstrapService:
        metadata:
          annotations:
            service.beta.kubernetes.io/openstack-internal-load-balancer: "true"
      perPodService:
        metadata:
          annotations:
            service.beta.kubernetes.io/openstack-internal-load-balancer: "true"
    # ...
```

You can specify different annotations for the bootstrap and the per-broker services.
After you specify these annotations, they will be passed by Strimzi to the Kubernetes services and the load balancers will be created accordingly.

## DNS annotations

**This feature will be available only since Strimzi 0.12.0.**

Many users are using some additional tools such as [ExternalDNS](https://github.com/kubernetes-incubator/external-dns) to automatically manage DNS records for their load balancers.
External DNS is using annotations on load balancer type services (and `Ingress` resources - more about that next time) to manage their DNS names.

> It supports many different DNs services such as Amazon AWS Route53, Google Cloud DNS, Azure DNS etc.

Reviewer: DNS

    overrides:
      bootstrap:
        dnsAnnotations:
          external-dns.alpha.kubernetes.io/hostname: kafka-bootstrap.mydomain.com.
          address: kafka-bootstrap.mydomain.com
      brokers:
      - broker: 0
        dnsAnnotations:
          external-dns.alpha.kubernetes.io/hostname: kafka-broker-0.mydomain.com.
          advertisedHost: kafka-broker-0.mydomain.com
      - broker: 1
        dnsAnnotations:
          external-dns.alpha.kubernetes.io/hostname: kafka-broker-1.mydomain.com.
          advertisedHost: kafka-broker-1.mydomain.com
      - broker: 2
        dnsAnnotations:
          external-dns.alpha.kubernetes.io/hostname: kafka-broker-2.mydomain.com.
          advertisedHost: kafka-broker-2.mydomain.com
# ...
```

# Pros and cons

The integration of load balancers into Kubernetes and OpenShift is very good and makes them very easy to use.
Strimzi can use the power of Kubernetes to provision them on many different public and private clouds.
Thanks to the TCP routing, you can freely decide whether you want to use TLS encryption or not.

The load balancers are also something what stands between the applications and the nodes of the Kubernetes cluster.

> It minimizes the attach surface and many admins would prefer load balancers over node ports for security reasons.

Reviewer: Attack?

Load balancers usually deliver a very good performance.
The typical load balancer is a service which runs outside of your cluster.

> So you do not need to be afraid of how much performance does it need, how much load will it put on your cluster and so on.

Reviewer 2: ```suggestion
So you do not need to be afraid of the resources it needs, how much load will it put on your cluster and so on.
```

However, there are some things to keep in mind:

> * In most cases Kubernetes will configure them to load balance across all cluster nodes.

Reviewer 2: ```suggestion
* In most cases, Kubernetes will configure them to load balance across all cluster nodes.
```

> So despite there being only one broker where the traffic will arrive at the end, different connections might be routed to the broker in through different cluster nodes and then forwarded through the `kube-proxy` to the right node where the Kafka broker actually runs.

Reviewer 2: ```suggestion
So despite there being only one broker where the traffic will ultimately arrive, different connections might be routed to that broker through different cluster nodes, being forwarded through the `kube-proxy` to the right node where the Kafka broker actually runs.
```

This would not happen for example with node ports, where the advertised address points directly to the node where the broker is running.

> * The load balancer it self is yet another service which the connection needs to go through.

Reviewer 2: ```suggestion
* The load balancer itself is yet another service which the connection needs to go through.
```

That might add a little bit of latency

Another aspect which you should consider is the price.

> In public clouds, the load balancers are normally not for free.

Reviewer 2: ```suggestion
In public clouds, the load balancers are normally not free.
```

> Usually you have to pay some fixed fee per instance which depends only on how long does the load balancer exist plus some fee for every transferred gigabyte.

Reviewer 2: ```suggestion
Usually you have to pay some fixed fee per instance which depends only on how long the load balancer exists plus some fee for every transferred gigabyte.
```

Strimzi always requires N+1 load balancers (where N is the number of brokers) - one for each broker + one for the bootstrapping.

> So you will always need multiple load balancers and the fees can easily add up.

Reviewer 2: ```suggestion
So you will always need multiple load balancers and the fees add up.
```

> Especially when you realize, that with Kafka in most cases the load balancer does't even balance the load because for most of the load balancers, there is only single broker behind them.

Reviewer 2: ```suggestion
And N of those load balancers don't even balance any load because there is only a single broker behind them.
```
